File: trip-kit/src/main/java/com/skedgo/android/tripkit/TripUpdaterImpl.java
package com.skedgo.android.tripkit;

import android.content.res.Resources;
import android.support.annotation.NonNull;
import android.text.TextUtils;

import com.skedgo.android.common.model.RoutingResponse;
import com.skedgo.android.common.model.Trip;
import com.skedgo.android.common.model.TripGroup;

import org.apache.commons.collections4.CollectionUtils;

import java.util.List;

import rx.Observable;
import rx.functions.Func1;
import rx.schedulers.Schedulers;

final class TripUpdaterImpl implements TripUpdater {
  private final Resources resources;
  private final TripUpdateApi api;
  private final String apiVersion;

  TripUpdaterImpl(
      Resources resources,
      TripUpdateApi api,
      String apiVersion) {
    this.resources = resources;
    this.api = api;
    this.apiVersion = apiVersion;
  }

  @NonNull @Override
  public Observable<Trip> getUpdateAsync(@NonNull Trip trip) {
    if (TextUtils.isEmpty(trip.getUpdateURL())) {
      return Observable.empty();
    }

    return api.fetchUpdateAsync(trip.getUpdateURL(), apiVersion)
        .map(new Func1<RoutingResponse, List<TripGroup>>() {
          @Override public List<TripGroup> call(RoutingResponse response) {
            response.processRawData(resources, GsonProvider.get());
            return response.getTripGroupList();
          }
        })
        .filter(new Func1<List<TripGroup>, Boolean>() {
          @Override public Boolean call(List<TripGroup> groups) {
            return CollectionUtils.isNotEmpty(groups);
          }
        })
        .map(new Func1<List<TripGroup>, Trip>() {
          @Override public Trip call(List<TripGroup> groups) {
            return groups.get(0).getDisplayTrip();
          }
        })
        .filter(new Func1<Trip, Boolean>() {
          @Override public Boolean call(Trip trip) {
            return trip != null;
          }
        })
        .subscribeOn(Schedulers.io());
  }
}

File: TripKitAndroid/src/main/java/com/skedgo/tripkit/TripUpdaterImpl.kt
package com.skedgo.tripkit

import android.content.res.Resources
import com.google.gson.Gson
import com.haroldadmin.cnradapter.NetworkResponse
import com.skedgo.tripkit.routing.Trip
import com.skedgo.tripkit.routing.TripGroup
import io.reactivex.Observable
import io.reactivex.schedulers.Schedulers
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.filter
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.flow.map

internal class TripUpdaterImpl(
    private val resources: Resources,
    private val api: TripUpdateApi,
    private val gson: Gson
) : com.skedgo.tripkit.TripUpdater {

    override fun getUpdateAsync(tripUrl: String): Observable<Trip> {
        val trip = Trip()
        trip.updateURL = tripUrl
        return getUpdateAsync(trip)
    }

    override fun getUpdateAsyncFlow(tripUrl: String): Flow<Trip> {
        val trip = Trip()
        trip.updateURL = tripUrl
        return getUpdateAsyncFlow(trip)
    }

    override fun getUpdateAsyncFlow(trip: Trip): Flow<Trip> {
        val updateURL = trip.updateURL
        if (updateURL.isNullOrEmpty()) {
            return flow { throw RuntimeException("Empty updateURL") }
        }
        return flow {
            val response = api.fetchUpdateAsyncKt(updateURL)
            when (response) {
                is NetworkResponse.Success -> {
                    response.body.processRawData(resources, gson)
                    emit(response.body.tripGroupList ?: listOf<TripGroup>())
                }

                else -> throw response.toException()
            }
        }.filter { it.isNotEmpty() }
            .filter { it.firstOrNull()?.displayTrip != null }
            .map { it.first().displayTrip!! }

    }

    override fun getUpdateAsync(trip: Trip): Observable<Trip> {
        val updateURL = trip.updateURL
        return if (updateURL.isNullOrEmpty()) {
            Observable.empty()
        } else {
            api.fetchUpdateAsync(updateURL)
                .map { response ->
                    response.processRawData(resources, gson)
                    response.tripGroupList ?: listOf<TripGroup>()
                }
                .filter {
                    it.isNotEmpty()
                }
                .filter { it.first().displayTrip != null }
                .map {
                    it.first().displayTrip!!
                }
                .subscribeOn(Schedulers.io())
        }
    }

    override fun tripSubscription(url: String): Observable<Any> =
        api.tripSubscription(url).subscribeOn(Schedulers.io())
}